### PR TITLE
update minimum versions

### PR DESCRIPTION
**Pull request by default-writer**
moving towards to minimum framework versions:

.NET 4.7.2
.NET Standard2.0
.NET Core 2.1

### CODE REVIEW

**Comment by jorive** [comment]
@hack2root Thanks for your contribution.
There are a lot of changes here. At first glance:
1. Refactoring work:
  * Remove `private` keyword.
  * Move open curly braces to a new line
  * use short notation `=>` on methods and properties
  * Sorted property names and fields
2. Moved the MarkdownLog code locally to remove the dependency in myget
3. Support netcoreapp2.1 and net471 (drop netstandard1.6, netcoreapp1.1, netcoreapp2.0)
4. Added code analysis
5. Removed the beta from the version
6. Moved to a newer version of xUnit
7. Cleaned up some Linq queries

It will take me couple of days before I go in details through all the changes. Did I miss something? What is the motivation for these changes?

File: dotnet-install.cmd [hunk 1]
```diff
@@ -11,7 +11,7 @@ setlocal
     exit /b 1
   )
 
-  set DotNet_Path=%~dp0tools\dotnet\%DotNet_Version%
+  set DotNet_Path=%~dp0packages\dotnet\%DotNet_Version%
```
Comment (jorive) [inline dotnet-install.cmd]: >packages [](start = 23, length = 8)

Please keep the DotNet Cli under the tools. Packages folder should be used for restored packages.

File: BuildCliComponents.sh [hunk 2]
```diff
@@ -1,58 +0,0 @@
-#!/bin/bash
```
Comment (jorive) [inline BuildCliComponents.sh]: Why delete this file? It might need to be updated, but we need an option for non-Windows platforms.

File: BuildForPublication.cmd [hunk 3]
```diff
@@ -1,45 +0,0 @@
-@echo off
```
Comment (jorive) [inline BuildForPublication.cmd]: Why delete this file?

File: CiBuild.cmd [hunk 4]
```diff
@@ -1,31 +0,0 @@
-@if not defined _echo echo off
```
Comment (jorive) [inline CiBuild.cmd]: This should not be deleted, it's used on the `netci.groovy` (which might also need to be updated).

File: _MyGet.cmd [hunk 5]
```diff
@@ -0,0 +1,111 @@
+@echo off
```
Comment (jorive) [inline _MyGet.cmd]: This is a duplicated `build.cmd` file without running the tests. Could you please remove this, and add a `-skiptests` option to `build.cmd` instead?

File: build.cmd [hunk 6]
```diff
@@ -5,10 +5,10 @@
 setlocal enabledelayedexpansion
   set errorlevel=
   set BuildConfiguration=%~1
-  if "%BuildConfiguration%"=="" set BuildConfiguration=Debug
-
-  set VersionSuffix=%~2
-  if "%VersionSuffix%"=="" set VersionSuffix=beta-build0000
+  if "%BuildConfiguration%"=="" set BuildConfiguration=Release
```
Comment (jorive) [inline build.cmd]: 👍 good call!

File: build.cmd [hunk 7]
```diff
@@ -72,10 +72,10 @@ setlocal
   cd /d %~dp0tests\simpleharness
   call :dotnet_build || exit /b 1
 
-  for %%v in (netcoreapp1.0 netcoreapp1.1 netcoreapp2.0 net461) do (
+  for %%v in (netcoreapp2.1) do (
```
Comment (jorive) [inline build.cmd]: Could you please keep netcoreapp2.0 support?

File: build.cmd [hunk 8]
```diff
@@ -90,10 +90,10 @@ setlocal
   cd /d %~dp0tests\scenariobenchmark
   call :dotnet_build || exit /b 1
 
-  for %%v in (netcoreapp1.0 netcoreapp1.1 netcoreapp2.0 net461) do (
+  for %%v in (netcoreapp2.1) do (
```
Comment (jorive) [inline build.cmd]: Could you please keep netcoreapp2.0 support?

File: myget.ps1 [hunk 9]
```diff
@@ -0,0 +1 @@
+& .\_MyGet
```
Comment (jorive) [inline myget.ps1]: This file is not needed.

File: dotnet-install.cmd [hunk 10]
```diff
@@ -39,7 +39,7 @@ setlocal
   )
 
   echo Executing dotnet installer script "%DotNet_Path%\dotnet-install.ps1"
-  for %%v in (1.0.0 1.1.0 2.0.0) do (
+  for %%v in (2.1.300) do (
```
Comment (jorive) [inline dotnet-install.cmd]: I think we should keep 2.0.3 here.

**Comment by jorive** [comment]
Thanks for integrating code analysis into the project! :)

File: src/xunit.performance.api/EtwPerformanceMetricEvaluationContext.cs [hunk 11]
```diff
@@ -12,10 +12,24 @@
...
+
+        readonly TraceEventSource _traceEventSource;
+
+        int _currentIteration;
+
+        string _currentTestCase;
+
         public EtwPerformanceMetricEvaluationContext(
-            string logPath,
+                                                                    string logPath,
```
Comment (jorive) [inline src/xunit.performance.api/EtwPerformanceMetricEvaluationContext.cs]: >                                                         [](start = 12, length = 56)

nit: extra spaces.

File: src/xunit.performance.api/xunit.performance.api.csproj [hunk 12]
```diff
@@ -4,21 +4,35 @@
 
   <PropertyGroup>
     <AssemblyTitle>xunit.performance.api</AssemblyTitle>
-    <TargetFrameworks>netstandard1.6;net461</TargetFrameworks>
-    <TargetFrameworks Condition="'$(OS)' != 'Windows_NT'">netstandard1.6</TargetFrameworks>
+    <TargetFrameworks>netstandard2.0;netcoreapp2.1;net471</TargetFrameworks>
```
Comment (jorive) [inline src/xunit.performance.api/xunit.performance.api.csproj]: >netcoreapp2.1 [](start = 37, length = 13)

netcoreapp2.1 is not needed here.

File: src/xunit.performance.api/xunit.performance.api.csproj [hunk 13]
```diff
@@ -4,21 +4,35 @@
...
+    <LangVersion>latest</LangVersion>
+    <DebugType>full</DebugType>
+    <DebugSymbols>true</DebugSymbols>
+  </PropertyGroup>
+
+  <PropertyGroup Condition="'$(Configuration)|$(TargetFramework)|$(Platform)'=='Release|netstandard2.0|AnyCPU'">
+    <LangVersion>latest</LangVersion>
+    <DebugType>none</DebugType>
+    <DebugSymbols>false</DebugSymbols>
   </PropertyGroup>
```
Comment (jorive) [inline src/xunit.performance.api/xunit.performance.api.csproj]: This should be added to `common.props` instead.

File: src/xunit.performance.api/xunit.performance.api.csproj [hunk 14]
```diff
@@ -4,21 +4,35 @@
...
+    <Version>1.0.0.0</Version>
+  </PropertyGroup>
+
+  <PropertyGroup Condition="'$(Configuration)|$(TargetFramework)|$(Platform)'=='Debug|netstandard2.0|AnyCPU'">
+    <LangVersion>latest</LangVersion>
+    <DebugType>full</DebugType>
+    <DebugSymbols>true</DebugSymbols>
+  </PropertyGroup>
+
+  <PropertyGroup Condition="'$(Configuration)|$(TargetFramework)|$(Platform)'=='Release|netstandard2.0|AnyCPU'">
```
Comment (jorive) [inline src/xunit.performance.api/xunit.performance.api.csproj]: >Condition="'$(Configuration)|$(TargetFramework)|$(Platform)'=='Release|netstandard2.0|AnyCPU'" [](start = 17, length = 94)

The condition should be:
`Condition="'$(Configuration)'=='Release'"` (or Debug)

File: src/xunit.performance.api/EtwPerformanceMetricEvaluationContext.cs [hunk 15]
```diff
@@ -76,7 +90,7 @@ internal class EtwPerformanceMetricEvaluationContext : PerformanceMetricEvaluati
             // The current process was already running before the etl tracing even started.
             benchmarkParser.Source.Kernel.ProcessStart += delegate (ProcessTraceData args)
             {
-                if (_currentProcesses.Contains(args.ParentID))
+                if (_currentProcesses.Contains(args.ProcessID))
```
Comment (jorive) [inline src/xunit.performance.api/EtwPerformanceMetricEvaluationContext.cs]: >ProcessID [](start = 52, length = 9)

Why the change from `ParentID` to `ProcessID`?

File: src/xunit.performance.api/PerformanceMonitorCounter/BasePerformanceMonitorCounter.cs [hunk 16]
```diff
@@ -4,34 +4,35 @@
...
             }
         }
 
-        public bool IsValidPmc => _profileSourceInfoID > -1;
-
-        protected int ProfileSourceInfoID => _profileSourceInfoID;
-
-        private readonly int _interval;
-        private readonly int _profileSourceInfoID;
+        protected int ProfileSourceInfoID { get; private set; }
```
Comment (jorive) [inline src/xunit.performance.api/PerformanceMonitorCounter/BasePerformanceMonitorCounter.cs]: >private set; [](start = 49, length = 12)

`private set;` is not needed.

File: src/xunit.performance.api/PerformanceMonitorCounter/BranchMispredictionsPerformanceMonitorCounter.cs [hunk 17]
```diff
@@ -1,13 +1,12 @@
 ﻿namespace Microsoft.Xunit.Performance.Api
 {
-    internal sealed class BranchMispredictionsPerformanceMonitorCounter : IPerformanceMonitorCounter
+    sealed class BranchMispredictionsPerformanceMonitorCounter : IPerformanceMonitorCounter
     {
         public string DisplayName => "Branch Mispredictions";
 
+        public int Interval => 1000;
```
Comment (jorive) [inline src/xunit.performance.api/PerformanceMonitorCounter/BranchMispredictionsPerformanceMonitorCounter.cs]: nit: add an extra line.

File: src/xunit.performance.api/PerformanceMonitorCounter/CacheMissesPerformanceMonitorCounter.cs [hunk 18]
```diff
@@ -1,13 +1,12 @@
 ﻿namespace Microsoft.Xunit.Performance.Api
 {
-    internal sealed class CacheMissesPerformanceMonitorCounter : IPerformanceMonitorCounter
+    sealed class CacheMissesPerformanceMonitorCounter : IPerformanceMonitorCounter
     {
         public string DisplayName => "Cache Misses";
 
+        public int Interval => 1000;
```
Comment (jorive) [inline src/xunit.performance.api/PerformanceMonitorCounter/CacheMissesPerformanceMonitorCounter.cs]: nit: Add an extra line.

File: src/xunit.performance.api/Profilers/ETWOutputData.cs [hunk 19]
```diff
@@ -1,13 +1,11 @@
 ﻿namespace Microsoft.Xunit.Performance.Api
 {
-    internal struct ETWOutputData
+    struct ETWOutputData
     {
+        public string KernelFileName { get; set; }
```
Comment (jorive) [inline src/xunit.performance.api/Profilers/ETWOutputData.cs]: nit: Add an extra line.

File: src/xunit.performance.core/xunit.performance.core.csproj [hunk 20]
```diff
@@ -4,12 +4,26 @@
...
+    <LangVersion>latest</LangVersion>
+    <DebugType>full</DebugType>
+    <DebugSymbols>true</DebugSymbols>
+  </PropertyGroup>
+
+  <PropertyGroup Condition="'$(Configuration)|$(TargetFramework)|$(Platform)'=='Release|netstandard2.0|AnyCPU'">
+    <LangVersion>latest</LangVersion>
+    <DebugType>none</DebugType>
+    <DebugSymbols>false</DebugSymbols>
   </PropertyGroup>
```
Comment (jorive) [inline src/xunit.performance.core/xunit.performance.core.csproj]: Same as the previous csproj, common properties should be moved to the `common.props` file

File: src/xunit.performance.execution/xunit.performance.execution.csproj [hunk 21]
```diff
@@ -4,10 +4,24 @@
...
+    <LangVersion>latest</LangVersion>
+    <DebugType>full</DebugType>
+    <DebugSymbols>true</DebugSymbols>
+  </PropertyGroup>
+
+  <PropertyGroup Condition="'$(Configuration)|$(TargetFramework)|$(Platform)'=='Release|netstandard2.0|AnyCPU'">
+    <LangVersion>latest</LangVersion>
+    <DebugType>none</DebugType>
+    <DebugSymbols>false</DebugSymbols>
   </PropertyGroup>
```
Comment (jorive) [inline src/xunit.performance.execution/xunit.performance.execution.csproj]: Same as previous csproj comments.

File: src/xunit.performance.metrics/xunit.performance.metrics.csproj [hunk 22]
```diff
@@ -4,18 +4,33 @@
...
+    <LangVersion>latest</LangVersion>
+    <DebugType>none</DebugType>
+    <DebugSymbols>false</DebugSymbols>
+  </PropertyGroup>
+
+  <PropertyGroup Condition="'$(Configuration)|$(TargetFramework)|$(Platform)'=='Debug|netstandard2.0|AnyCPU'">
+    <LangVersion>latest</LangVersion>
+    <DebugType>full</DebugType>
+    <DebugSymbols>true</DebugSymbols>
   </PropertyGroup>
```
Comment (jorive) [inline src/xunit.performance.metrics/xunit.performance.metrics.csproj]: Same as previous csproj comments.

File: tests/scenariobenchmark/scenariobenchmark.csproj [hunk 23]
```diff
@@ -1,19 +1,56 @@
...
-    <TargetFrameworks>netcoreapp1.0;netcoreapp1.1;netcoreapp2.0;net461</TargetFrameworks>
+    <TargetFrameworks>
+      <!-- netstandard1.0; -->
+      <!-- netstandard1.1; -->
+      <!-- netstandard1.2; -->
+      <!-- netstandard1.3; -->
+      <!-- netstandard1.4; -->
+      <!-- netstandard1.5; -->
+      <!-- netstandard1.6; -->
+      netstandard2.0;
```
Comment (jorive) [inline tests/scenariobenchmark/scenariobenchmark.csproj]: it should be netcoreapp2.0

File: tests/scenariobenchmark/scenariobenchmark.csproj [hunk 24]
```diff
@@ -1,19 +1,56 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup Label="Configuration">
-    <TreatWarningsAsErrors>True</TreatWarningsAsErrors>
+    <TreatWarningsAsErrors>False</TreatWarningsAsErrors>
```
Comment (jorive) [inline tests/scenariobenchmark/scenariobenchmark.csproj]: >False [](start = 27, length = 5)

It should be `True`.

File: tests/scenariobenchmark/scenariobenchmark.csproj [hunk 25]
```diff
@@ -1,19 +1,56 @@
...
+    <DebugType>full</DebugType>
+    <DebugSymbols>true</DebugSymbols>
+  </PropertyGroup>
+
   <ItemGroup>
     <ProjectReference Include="..\..\src\xunit.performance.api\xunit.performance.api.csproj" />
     <ProjectReference Include="..\..\src\xunit.performance.core\xunit.performance.core.csproj" />
-    <ProjectReference Include="..\..\src\xunit.performance.execution\xunit.performance.execution.csproj" />
+    <!-- <ProjectReference Include="..\..\src\xunit.performance.execution\xunit.performance.execution.csproj" /> -->
     <ProjectReference Include="..\..\src\xunit.performance.metrics\xunit.performance.metrics.csproj" />
```
Comment (jorive) [inline tests/scenariobenchmark/scenariobenchmark.csproj]: I think you can remove these lines.

File: tests/simpleharness/simpleharness.csproj [hunk 26]
```diff
@@ -1,23 +1,60 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup Label="Configuration">
-    <TreatWarningsAsErrors>True</TreatWarningsAsErrors>
+    <TreatWarningsAsErrors>False</TreatWarningsAsErrors>
```
Comment (jorive) [inline tests/simpleharness/simpleharness.csproj]: >False [](start = 27, length = 5)

It should be `True`

File: tests/simpleharness/simpleharness.csproj [hunk 27]
```diff
@@ -1,23 +1,60 @@
...
-    <TargetFrameworks>netcoreapp1.0;netcoreapp1.1;netcoreapp2.0;net461</TargetFrameworks>
+    <TargetFrameworks>
+      <!-- netstandard1.0; -->
+      <!-- netstandard1.1; -->
+      <!-- netstandard1.2; -->
+      <!-- netstandard1.3; -->
+      <!-- netstandard1.4; -->
+      <!-- netstandard1.5; -->
+      <!-- netstandard1.6; -->
+      netstandard2.0;
```
Comment (jorive) [inline tests/simpleharness/simpleharness.csproj]: >netstandard2.0 [](start = 6, length = 14)

It should be `netcoreapp2.0`

File: tests/simpleharness/simpleharness.csproj [hunk 28]
```diff
@@ -1,23 +1,60 @@
...
   <ItemGroup>
     <None Remove="*.csv;*.etl;*.md;*.xml" />
   </ItemGroup>
 
   <ItemGroup>
     <ProjectReference Include="..\..\src\xunit.performance.api\xunit.performance.api.csproj" />
     <ProjectReference Include="..\..\src\xunit.performance.core\xunit.performance.core.csproj" />
-    <ProjectReference Include="..\..\src\xunit.performance.execution\xunit.performance.execution.csproj" />
+    <!--<ProjectReference Include="..\..\src\xunit.performance.execution\xunit.performance.execution.csproj" />-->
     <ProjectReference Include="..\..\src\xunit.performance.metrics\xunit.performance.metrics.csproj" />
```
Comment (jorive) [inline tests/simpleharness/simpleharness.csproj]: I think you can remove these lines.

File: tests/standalonesample/standalonesample.csproj [hunk 29]
```diff
@@ -1,20 +1,62 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup Label="Configuration">
-    <TreatWarningsAsErrors>True</TreatWarningsAsErrors>
+    <TreatWarningsAsErrors>False</TreatWarningsAsErrors>
```
Comment (jorive) [inline tests/standalonesample/standalonesample.csproj]: >False [](start = 27, length = 5)

It should be `True`

File: tests/standalonesample/standalonesample.csproj [hunk 30]
```diff
@@ -1,20 +1,62 @@
...
-    <TargetFrameworks>netcoreapp1.0;netcoreapp1.1;netcoreapp2.0;net461</TargetFrameworks>
+    <TargetFrameworks>
+      <!-- netstandard1.0; -->
+      <!-- netstandard1.1; -->
+      <!-- netstandard1.2; -->
+      <!-- netstandard1.3; -->
+      <!-- netstandard1.4; -->
+      <!-- netstandard1.5; -->
+      <!-- netstandard1.6; -->
+      netstandard2.0;
```
Comment (jorive) [inline tests/standalonesample/standalonesample.csproj]: >netstandard2.0 [](start = 6, length = 14)

It should be `netcoreapp2.0`

File: tests/standalonesample/standalonesample.csproj [hunk 31]
```diff
@@ -1,20 +1,62 @@
...
+    <PackageReference Include="Microsoft.Diagnostics.Tracing.TraceEvent" Version="2.0.19">
       <IncludeAssets>All</IncludeAssets>
     </PackageReference>
-    <PackageReference Include="xunit.performance.api" Version="1.0.0-beta-build0015" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\..\src\xunit.performance.api\xunit.performance.api.csproj" />
+    <ProjectReference Include="..\..\src\xunit.performance.core\xunit.performance.core.csproj" />
+    <!--<ProjectReference Include="..\..\src\xunit.performance.execution\xunit.performance.execution.csproj" />-->
```
Comment (jorive) [inline tests/standalonesample/standalonesample.csproj]: I think you can remove these lines.

**Comment by jorive** [comment]
@hack2root the build failed because the script run by the CI was deleted/renamed.

**Comment by brianrob** [comment]
Closing prior to archive.
